mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -1306,4 +1310,30 @@ object ALS extends DefaultParamsReadable[ALS] with Logging {
    * satisfies this requirement, we simply use a type alias here.
    */
   private[recommendation] type ALSPartitioner = org.apache.spark.HashPartitioner
+
+  /**
+   * Private function to clean up all of the shuffles files from the dependencies and their parents.
+   */
+  private[spark] def cleanShuffleDependencies[T](sc: SparkContext, deps: Seq[Dependency[_]],
+      blocking: Boolean = false): Unit = {
+    // If there is no reference tracking we skip clean up.
+    if (sc.cleaner.isEmpty) {
+      return
+    }
+    val cleaner = sc.cleaner.get
+    /**
+     * Clean the shuffles & all of its parents.
+     */
+    def cleanEagerly(dep: Dependency[_]): Unit = {
```

Comment:
Oh right, I see that now. OK. Are we worried the dependency graph will be super deep? probably not right, since chained shuffle dependencies woudl be rare?